pubtools/_ami/tasks/push.py

```diff
@@ -452,12 +458,11 @@ def add_args(self):
 
         group.add_argument(
             "--snapshot-account-ids",
-            help="Comma separated list of account ids to give snapshot "
-            "creation permissions to if a new snapshot is created as "
-            "part of the image push.",
-            action=SplitAndExtend,
-            split_on=",",
-            remove_duplicates=True,
+            help="JSON string mapping region to a list of account ids to give "
+            "snapshot  creation permissions to if a new snapshot is created "
+            "as part of the image push.",
+            type=json.loads,
```

Comment:
This looks like a backwards-incompatible change, i.e. passing the previous argument format will cause a crash.
Is there some way to make it backwards-compatible?

By making it backwards-incompatible a deployment problem is created.  Looking at the code which provides this argument, it's copied straight out of a target setting.  That creates a situation where the target setting has to be updated exactly at the same time as we upgrade pubtools-ami to this new format (and then rolled back again if we have to roll back the upgrade).  It might not sound that difficult but in practice, making sure that happens across all environments with the correct timing is error-prone.

If it's technically feasible to do so, and unless there is some reason that you know of that the backwards-incompatible change wouldn't matter (e.g. you know that nobody is using this feature yet), I think the change should be done in a backwards-compatible way.  i.e. accept both the new JSON format and the old comma-separated.  Some of the code mentions a `"default"` region, so maybe a value of `"a,b,c"` can be silently converted to `{"default": ["a", "b", "c"]}` ?

Reply:
I took a quick look at the target settings and "snapshot_account_ids" are not defined in any of the AWS targets. Furthermore, this change was released in v1.2.0 but the latest on prod is v1.1.0. So I believe this impl is not being used currently and the default ids from initial implementation for snapshot id are still in use.
However, as mentioned it should be ensured that target settings are updated by the time this change is deployed..

Reply:
OK, if the old format is not currently used, in the interest of avoiding unnecessary work I'll withdraw this objection.

This is not really a good practice going forward though, it's intended and important for these projects to have a stable interface.  It makes everyone's life harder if we add arguments and then change our minds (multiple times?) over a few months on the format.  We even claim to follow SemVer in https://github.com/release-engineering/pubtools-ami/blob/main/CHANGELOG.md which means the next release technically should be 2.0.0 if we don't maintain backwards-compatibility.